internal/login/login.go

```diff
@@ -1,18 +1,225 @@
+	}
+
+	block, err := aes.NewCipher(secret)
+	if err != nil {
+		return "", fmt.Errorf("%s: %w", decryptionErrorMsg, err)
+	}
+
+	aesgcm, err := cipher.NewGCM(block)
+	if err != nil {
+		return "", fmt.Errorf("%s: %w", decryptionErrorMsg, err)
+	}
+
+	decryptedAccessToken, err := aesgcm.Open(nil, nonce, decodedAccessToken, nil)
+	if err != nil {
+		return "", fmt.Errorf("%s: %w", decryptionErrorMsg, err)
+	}
+
+	return string(decryptedAccessToken), nil
+}
+
+func pollForAccessToken(url string) (AccessTokenResponse, error) {
+	var accessTokenResponse AccessTokenResponse
+
+	// We fully control the url here, so it's safe to perform the request and ignore the G107 gosec rule.
+	// TODO: Move to OpenAPI-generated http client once we reach v1 on API schema.
+	resp, err := http.Get(url) //nolint:gosec
+	if err != nil {
+		return accessTokenResponse, fmt.Errorf("cannot fetch access token: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode == http.StatusNotFound {
+		retryAfterSeconds, err := strconv.Atoi(resp.Header.Get("Retry-After"))
+		if err != nil {
+			retryAfterSeconds = defaultRetryAfterSeconds
+		}
+		time.Sleep(time.Duration(retryAfterSeconds) * time.Second)
+		return pollForAccessToken(url)
+	}
+
+	if resp.StatusCode == http.StatusOK {
+		body, err := io.ReadAll(resp.Body)
+
+		if err != nil {
+			return accessTokenResponse, fmt.Errorf("cannot read access token response body: %w", err)
+		}
+
+		if err := json.Unmarshal(body, &accessTokenResponse); err != nil {
+			return accessTokenResponse, fmt.Errorf("cannot unmarshal access token response: %w", err)
+		}
+
+		return accessTokenResponse, nil
+	}
+
+	return accessTokenResponse, fmt.Errorf("HTTP %s: cannot retrieve access token", resp.Status)
+}
+
+func Run(stdin *os.File, params RunParams) error {
+	if params.Token != "" {
+		err := utils.SaveAccessToken(params.Token, params.Fsys)
+		if err != nil {
+			return fmt.Errorf("cannot save provided token: %w", err)
+		}
+		fmt.Println(loggedInMsg)
+		return nil
+	}
+
+	if params.OpenBrowser {
+		fmt.Print("Hello from ", utils.Aqua("Supabase"), "! Press ", utils.Aqua("Enter"), " to open browser and login automatically.\n")
+		fmt.Scanln()
+	}
+
+	sessionId := uuid.New().String()
+	tokenName := params.Name
+	curve := ecdh.P256()
+	privateKey, err := curve.GenerateKey(rand.Reader)
+	if err != nil {
+		return fmt.Errorf("cannot generate encryption key: %w", err)
+	}
+	publicKey := privateKey.PublicKey()
+	encodedPublicKey := hex.EncodeToString(publicKey.Bytes())
+
+	createLoginSessionPath := "/cli/login"
+	createLoginSessionQuery := "?session_id=" + sessionId + "&token_name=" + tokenName + "&public_key=" + encodedPublicKey
+	createLoginSessionUrl := utils.GetSupabaseDashboardURL() + createLoginSessionPath + createLoginSessionQuery
+
+	if params.OpenBrowser {
+		fmt.Printf("Here is your login link in case browser did not open: %s\n\n", utils.Bold(createLoginSessionUrl))
+
+		openCmd := exec.Command("open", createLoginSessionUrl)
```

Comment:
```suggestion
		openCmd := exec.CommandContext(ctx,
"open", createLoginSessionUrl)
```

Use context to allow cancellation.